### PR TITLE
feat: Support CUDA graphs for EAGLE3

### PR DESCRIPTION
**Pull request by mikeiovine**
Support CUDA graphs for the EAGLE-3 spec decode. Also contains fixes for loading eagle 3 weights for llama3 70B models (previously only tested for 8b).

The graphs significantly improve the performance. However, we still have a lot of work to do to eliminate the host overheads.

### CODE REVIEW

File: tensorrt_llm/_torch/pyexecutor/model_engine.py [hunk 1]
```diff
@@ -1759,3 +1773,23 @@ def _init_userbuffers(self, hidden_size):
...
+            loader(target_model)
+
+    @contextlib.contextmanager
+    def disable_cuda_graph(self):
+        _run_cuda_graphs = self._run_cuda_graphs
+        self._run_cuda_graphs = False
+        try:
+            yield
+        finally:
+            self._run_cuda_graphs = _run_cuda_graphs
```
Comment (mikeiovine) [inline tensorrt_llm/_torch/pyexecutor/model_engine.py]: Forgot to deduplicate with an identical function above; will do in next update.

**Comment by mikeiovine** [comment]
Putting this out to get early feedback and ideas. I'm not really happy with the design right now. Specifically, passing the states between the target and draft models gets pretty complicated when CUDA graphs enter the picture.

File: tensorrt_llm/_torch/attention_backend/interface.py [hunk 2]
```diff
@@ -160,7 +160,16 @@ def seq_lens(self, value: Optional[torch.Tensor]):
...
-            self._seq_lens_cuda = self._seq_lens.cuda(non_blocking=True)
+
+            if self.is_cuda_graph and self._seq_lens_cuda is not None:
+                # Very important: do not reallocate if we are using CUDA graphs.
+                # This copy is safe because the batch size is guaranteed to not
+                # change in the CUDA graph case. The seqlens can change if we
+                # are doing spec decode.
+                self._seq_lens_cuda.copy_(self._seq_lens)
+            else:
+                self._seq_lens_cuda = self._seq_lens.cuda(non_blocking=True)
```
Comment (hlu1) [inline tensorrt_llm/_torch/attention_backend/interface.py:174]: This part is a bit hard to follow. When will this branch get hit?

How are these two lines different:
```
self._seq_lens_cuda.copy_(self._seq_lens)
self._seq_lens_cuda = self._seq_lens.cuda(non_blocking=True)
```
On first look, they're both copying from `_seq_lens` to `_seq_lens_cuda`

File: tensorrt_llm/_torch/pyexecutor/model_engine.py [hunk 3]
```diff
@@ -1113,13 +1127,12 @@ def _prepare_tp_inputs(
...
-            # and have static batch size, so the seqlens never change.
-            # Note that it's important to not free the seq_lens_cuda
-            # buffer once the graph has been captured also - this will invalidate
-            # the graph and force an expensive recapture.
+        if not attn_metadata.is_cuda_graph or is_spec_decode:
+            # Usually, we don't need to update seq_lens when using CUDA graphs.
+            # This is because CUDA graphs are only used for pure decoding batches.
+            # If we're doing spec decode, however, we can have variable seqlens (up
+            # to max_num_draft_tokens per request). The buffers inside the CUDA graph
+            # runner are padded to handle this.
```
Comment (hlu1) [inline tensorrt_llm/_torch/pyexecutor/model_engine.py]: Can you add a reference to where it's padded?

File: tensorrt_llm/_torch/models/modeling_llama.py [hunk 4]
```diff
@@ -743,7 +761,33 @@ def forward(
...
             new_weights[new_k] = v
 
         super().load_weights(new_weights)
+
+    def load_weights_from_target_model(self,
+                                       target_model: torch.nn.Module) -> None:
+        if self.model.embed_tokens is None:
+            self.model.embed_tokens = target_model.model.embed_tokens
+
+    def spec_decode_extra_input_info(self, batch_size: int,
```
Comment (hlu1) [inline tensorrt_llm/_torch/models/modeling_llama.py]:
```suggestion
    def spec_decode_inputs(self, batch_size: int,
```

File: tensorrt_llm/_torch/pyexecutor/model_engine.py [hunk 5]
```diff
@@ -527,7 +538,9 @@ def no_cuda_graph():
                         logger.info(f"Run autotuning warmup for batch size={1}")
                         self.forward(batch,
                                      new_tensors_device=None,
-                                     resource_manager=resource_manager)
+                                     resource_manager=resource_manager,
+                                     extra_model_inputs=_create_extra_inputs(
+                                         1, num_tokens))
```
Comment (hlu1) [inline tensorrt_llm/_torch/pyexecutor/model_engine.py]: Why bs = 1 here?

File: tensorrt_llm/_torch/pyexecutor/model_engine.py [hunk 6]
```diff
@@ -553,7 +566,8 @@ def no_cuda_graph():
                 logger.info(f"Run warmup for batch size={bs}")
                 self.forward(batch,
                              new_tensors_device=None,
-                             resource_manager=resource_manager)
+                             resource_manager=resource_manager,
+                             extra_model_inputs=_create_extra_inputs(bs, 1))
```
Comment (hlu1) [inline tensorrt_llm/_torch/pyexecutor/model_engine.py:576]: Why num_tokens = 1 here?

**Review by hlu1 — COMMENTED** [review]
Mostly nits.

File: tensorrt_llm/_torch/pyexecutor/model_engine.py (hunk 5 above)
Comment (mikeiovine) [reply tensorrt_llm/_torch/pyexecutor/model_engine.py]: This one is doing a warmup with a single request with `num_tokens`. See L532.

File: tensorrt_llm/_torch/pyexecutor/model_engine.py (hunk 6 above)
Comment (mikeiovine) [reply tensorrt_llm/_torch/pyexecutor/model_engine.py:576]: That should probably be renamed `num_tokens_per_batch`. This one is just doing decoding request warmups. So 1 token each.

**Comment by mikeiovine** [comment]
Had some discussion offline with @hlu1 about how to make it cleaner:

1. `extra_model_inputs` is hard to extend. A single `class ModelInput` that all models consume would allow us to easily add new features to other models.
2. At the same time, `spec_decode_extra_input_info` can be generalized. ~~We can rename it to `get_input_shapes`, and it can return a dict (`str -> (shape, dtype)`). It can have some sensible defaults to avoid the burden of having to implement this for more standard models.~~ Decided to just rename it `get_warmup_extra_inputs`.

I think (2) should definitely be done now. (1) is a pretty big refactor and would probably have to be done in followups.

File: tensorrt_llm/_torch/attention_backend/interface.py (hunk 2 above)
Comment (mikeiovine) [reply tensorrt_llm/_torch/attention_backend/interface.py:174]: I think the comment explains it. The difference is that the second version creates a new tensor while the first copies into an existing one. In eager mode, you have to create a new one because the batch size might have changed. In CUDA graph mode, the batch size is static. You are forced to reuse the same tensor to avoid invalidating the graph.

File: tensorrt_llm/_torch/pyexecutor/cuda_graph_runner.py [hunk 7]
```diff
@@ -80,6 +99,12 @@ def capture(
             "inputs_embeds": None,
             "spec_metadata": self.spec_metadata,
         }
+        if extra_model_inputs is not None:
+            for key, tensor in extra_model_inputs.items():
+                new_tensor = tensor.clone()
```
Comment (hlu1) [inline tensorrt_llm/_torch/pyexecutor/cuda_graph_runner.py:104]: >  `extra_model_inputs: Optional[Dict[str, Any]]`

Here you're assuming that it's a torch.Tensor. Should you change the typehints? For example, line 131 down blow has `extra_model_inputs: Optional[Dict[str, torch.Tensor]]`

File: tensorrt_llm/_torch/pyexecutor/model_engine.py (hunk 6 above)
Comment (hlu1) [reply tensorrt_llm/_torch/pyexecutor/model_engine.py:576]: Could you add the renaming? It's a simple change but makes it easier to understand.

File: tensorrt_llm/_torch/pyexecutor/model_engine.py [hunk 8]
```diff
@@ -1759,3 +1776,14 @@ def _init_userbuffers(self, hidden_size):
                                           self.mapping.gpus_per_node,
                                           hidden_size * self.max_num_tokens * 2)
         return True
+
+    def load_weights_from_target_model(self,
```
Comment (hlu1) [inline tensorrt_llm/_torch/pyexecutor/model_engine.py:1776]: Where is the method called?

File: tensorrt_llm/_torch/pyexecutor/py_executor.py [hunk 9]
```diff
@@ -1684,14 +1684,19 @@ def _prepare_draft_tokens(self, scheduled_requests: ScheduledRequests):
...
+                hidden_states = hidden_states.to(
+                    self.draft_model_engine.model.model.dtype)
+                # Hack for eagle3. We might need to run a matmul to reduce
+                # the dimensionality of the hidden states on the first pass
+                # through the draft model. Shape dependent control flow will
+                # not work with CUDA graphs. So we just do it here.
+                expected_hidden_size = self.draft_model_engine.model.model.hidden_size
+                fc = self.draft_model_engine.model.model.fc
+                if hidden_states.shape[-1] != expected_hidden_size:
+                    hidden_states = fc(hidden_states)
```
Comment (hlu1) [inline tensorrt_llm/_torch/pyexecutor/py_executor.py]: Can you create a function for this in the eagle model and call that function here? It's weird to have model code here.

File: tensorrt_llm/_torch/pyexecutor/model_engine.py (hunk 8 above)
Comment (mikeiovine) [reply tensorrt_llm/_torch/pyexecutor/model_engine.py:1776]: See `py_executor_creator.py:L98`. I'm not a huge fan of doing this but unfortunately the eagle checkpoint format that we're using sometimes is missing `embed_tokens` because it's meant to be shared with the target model.

File: tensorrt_llm/_torch/pyexecutor/cuda_graph_runner.py (hunk 7 above)
Comment (mikeiovine) [reply tensorrt_llm/_torch/pyexecutor/cuda_graph_runner.py:104]: Updated, thanks. In a previous version this was more general.

File: tensorrt_llm/_torch/pyexecutor/model_engine.py (hunk 6 above)
Comment (mikeiovine) [reply tensorrt_llm/_torch/pyexecutor/model_engine.py:576]: Yep, I think it's already in there.

File: tensorrt_llm/_torch/models/modeling_llama.py [hunk 10]
```diff
@@ -652,32 +658,50 @@ def __init__(self, model_config: ModelConfig[LlamaConfig]) -> None:
...
+
     def forward(
         self,
         attn_metadata: AttentionMetadata,
-        embed_tokens: torch.nn.Module,
         input_ids: Optional[torch.LongTensor] = None,
         position_ids: Optional[torch.LongTensor] = None,
         inputs_embeds: Optional[torch.FloatTensor] = None,
         spec_metadata: Optional[SpecMetadata] = None,
         hidden_states: Optional[torch.Tensor] = None,
```
Comment (hlu1) [inline tensorrt_llm/_torch/models/modeling_llama.py:782]: The type is incorrect. It's not optional. Is it a single tensor now?

File: tensorrt_llm/_torch/models/modeling_llama.py [hunk 11]
```diff
@@ -686,7 +710,7 @@ def forward(
         hidden_states, hidden_states_to_save = self.norm(
             hidden_states, residual)
         assert isinstance(spec_metadata, Eagle3SpecMetadata)
-        spec_metadata.hidden_states.append(hidden_states_to_save)
+        spec_metadata.maybe_capture_hidden_states(1, hidden_states_to_save)
```
Comment (hlu1) [inline tensorrt_llm/_torch/models/modeling_llama.py:808]: Why is the layer_id set to 1 here? Overall I don't understand how you pass the intermediate target model states to the eagle draft model. Like where and how you pass them?

File: tensorrt_llm/_torch/models/modeling_llama.py [hunk 12]
```diff
@@ -743,7 +758,33 @@ def forward(
...
+                                num_tokens: int) -> Dict[str, Any]:
+
+        hidden_states = torch.empty(batch_size * num_tokens,
+                                    self.model.hidden_size_in,
+                                    dtype=self.model.dtype,
+                                    device='cuda')
+
+        return {'hidden_states': hidden_states}
+
+    def apply_eagle3_fc(self, hidden_states: torch.Tensor) -> torch.Tensor:
```
Comment (hlu1) [inline tensorrt_llm/_torch/models/modeling_llama.py:878]: Can you add some comments on why this has to be a separate function and where it's called? Similar to the comments you add to inside `_prepare_draft_tokens`.

File: tensorrt_llm/_torch/speculative/eagle3.py [hunk 13]
```diff
@@ -86,7 +106,12 @@ def get_hidden_states(
             hidden_states_gather_ids.extend(
                 req_id_to_gather_ids[req.py_request_id])
 
-        return [h[hidden_states_gather_ids] for h in self.hidden_states]
+        if len(self.hidden_states) == 1:
+            return self.hidden_states[0][hidden_states_gather_ids]
+        else:
+            return torch.cat(
```
Comment (hlu1) [inline tensorrt_llm/_torch/speculative/eagle3.py:112]: Why is the cat moved here?

**Review by hlu1 — APPROVED** [review]
Approve to unblock.

File: tensorrt_llm/_torch/models/modeling_llama.py (hunk 10 above)
Comment (mikeiovine) [reply tensorrt_llm/_torch/models/modeling_llama.py:782]: Yeah. We have an assertion that it's not none. It has to have this annotation to match the base class unfortunately.

File: tensorrt_llm/_torch/models/modeling_llama.py (hunk 11 above)
Comment (mikeiovine) [reply tensorrt_llm/_torch/models/modeling_llama.py:808]: You copy the hidden states into the spec metadata for a specified set of layer outputs. We collect the hidden states from the spec metadata later in the executor. 

~As for the 1, it's because the eagle specification is weird. We already have a decoder layer with layer ID zero. But we don't want to capture the outputs from it when we run the draft model, we want to capture the outputs after an additional norm. So give it the label 1 instead of zero to avoid capturing the wrong states.~

Edit: was pretty much wrong about the above. We're just capturing `hidden_states + residual`, not the normed output. Fixing in follow up...

File: tensorrt_llm/_torch/speculative/eagle3.py (hunk 13 above)
Comment (mikeiovine) [reply tensorrt_llm/_torch/speculative/eagle3.py:112]: If it was inside the model, we would have shape dependent control flow, which will break CUDA graphs.

**Comment by mikeiovine** [comment]
Running CI one more time before merging as it has been a while since my last rebase.

**Comment by mikeiovine** [comment]
/bot skip --comment  "Pipeline passed before last rebase"

File: tensorrt_llm/_torch/speculative/eagle3.py (hunk 13 above)
Comment (hlu1) [reply tensorrt_llm/_torch/speculative/eagle3.py:112]: Could you add a comment here please in your next eagle PR?

File: tensorrt_llm/_torch/models/modeling_llama.py (hunk 11 above)
Comment (hlu1) [reply tensorrt_llm/_torch/models/modeling_llama.py:808]: Could you add these to the code as comments? It's hard to infer all this from the code.
